recipes/brpc/all/CMakeLists.txt

```diff
@@ -0,0 +1,23 @@
+cmake_minimum_required(VERSION 2.8.11)
+project(cmake_wrapper)
+
+include(conanbuildinfo.cmake)
+conan_basic_setup()
+
+# disable picking up system-installed libraries with find_library
+set(CMAKE_FIND_USE_CMAKE_SYSTEM_PATH FALSE)
+
+# we need to set the executable manually since that seems to be missing in
+# the generated FindProtobuf.cmake in the protobuf conan package
+find_program(PROTOBUF_PROTOC_EXECUTABLE NAMES protoc)
```

Comment:
This won't work currently on Macos when protbuf is built as a shared library.
Maybe add a FIXME with more info for easier grep'ing later.

Reply:
Why is that? (Thanks for working on the protobuf recipe btw)

Reply:
Apply SIP (system integration protection): what this means here is that the `DYLD_LIBRARY_PATH` string **must** be prepended to the EXECUTABLE. The path of the shared library being in the `DYLD_LIBRARY_PATH` environment variable is not enough.
See my (merged) protobuf recipe for my agony.

Reply:
I tried the recipe on a Mac and I think I ran into this. I have a few other fixes as well to add for MacOS.

That said, the latest commit did result in some MacOS packages even with shared. I dumped the full sorted list here: https://gist.github.com/jjkoshy/47018ef1d02528f2dd6d66501ba97afd

I'm not sure how those worked.

Reply:
It will run into problems when *protobuf* is built as a shared library, because the `protoc` binary needs the `libprotobuf.dylib` library.

Reply:
Yeah I realized what you meant right soon after I wrote the previous comment :) - I'll try and patch that as well. If that's harder than I anticipate I'll defer it to a FIXME.

Reply:
I'm fine with starting this out with limited MacOS/Windows support but I'm a bit surprised with the latest CI run.

With the latest changes, I was able to generate packages with `-o protobuf=shared`. The [latest CI run](https://github.com/conan-io/conan-center-index/pull/2370#issuecomment-671107597) generated 130 packages which is the same as the [previous CI run](https://github.com/conan-io/conan-center-index/pull/2370#issuecomment-669502941). I was expecting to see more.

Is there a way I can see what configurations were _not_ generated and why?

Reply:
Hmm.. ok now I'm confused. [The comment](https://github.com/conan-io/conan-center-index/pull/2370#issuecomment-671107597) says "generated 130 packages" but when I actually look at the logs I see this in all the `create_stderr.txt`:

```
ERROR: brpc/0.9.6: Invalid configuration: brpc requires protobuf:with_zlib=True
```

i.e., no package were actually generated.

Is this related to https://github.com/conan-io/conan-center-index/pull/2370#discussion_r465983515 ?

Does the CI only attempt to generate packages for the _default_ options of the recipe and its dependencies?

Reply:
@danimtb @uilianries 
Something went wrong with CI.

Reply:
Not sure about the commit (CI job) related to this comment, but I will guess.... this recipe raises `ConanInvalidConfiguration` from the `build()` method, CCI doesn't propagate the error for that exception, it is a successful job, so it counts it as another package although the package is not generated at all. You get 130 packages, but some of them _succeeded_ with that exception.

This is something we need to change so the information is not misleading.

---

With this, we have a problem because we need to actually run the `build` in order to know that certain configuration will raise a `ConanInvalidConfiguration`. The problem is that with this in mind we can no longer iterate all the profiles, compute the package-ids and remove configurations that result in a different package-id... and be sure that the rest will be successful.

Reply:
Hi @jgsogo thanks for the response. I think all 130 packages are empty.

> This is something we need to change so the information is not misleading.

Is there an issue filed for that?

> The problem is that with this in mind we can no longer iterate all the profiles, compute the package-ids and remove configurations that result in a different package-id... and be sure that the rest will be successful.

I'm not sure what you mean by "remove configurations that result in a different package-id". Either way, I'm unclear on how to move forward with this. @madebr would this be a scenario where it would be appropriate to pin `protobuf:with_zlib` to be true? (Ref: https://github.com/conan-io/conan-center-index/pull/2370#discussion_r465745218)

Reply:
@jjkoshy setting `protobuf:with_zlib` will probably result in a CI error about that package not being available.

Reply:
We have merged a PR that will report the number of packages built and the number of packages skipped because of a `ConanInvalidConfiguration` raised from the `build()` method. Output will be something like this: 

```
`zlib/1.2.2`: Generated 2 packages (+ 1 invalid config from `build()`)
```

instead of the `Generated 3 packages` we got before. Probably generating 0 packages should be an error (it is a reported issue, it should be really easy to implement now that we are counting packages properly).

---

> @jjkoshy setting `protobuf:with_zlib` will probably result in a CI error about that package not being available.

Exploding options (besides `shared` or `fPIC`) is something we are not doing because of combinatorial explosion and the number of packages to be generated 💣 ... but after https://github.com/conan-io/conan-center-index/issues/2408#issuecomment-669097060 we are starting to think how to add this feature in a controlled manner. Probably this is something to add to the `config.yml`, but we need to control who can add options there.

Here I think all 130 are packages are empty because of this reason 😢 

I need to ask you for a little more time, we need to figure out how to implement it

Reply:
Thanks for the update. In the meantime I thought I would just try and circumvent by voiding the parts of brpc code which depends on zlib when `with_zlib` is set to false. It's a hack, and not sure if it will help. An alternative as @madebr mentioned is to consider revisiting the `with_zlib` default of protobuf.